File: CSE241 - Object Oriented Programming/studies/cpp/ch15-Polymorphism-and-Virtual-Functions/employee/commissionEmployee.cpp
#include <iostream>
#include <string>
#include "commissionEmployee.h"

using std::string;
using std::cerr;
using std::cout;

namespace GTUEmployee {
    CommissionEmployee::CommissionEmployee (const string & firstName, const string & lastName, const string & ssn, double grossSales, double commissionRate)
    : Employee(firstName, lastName, ssn) {
        setGrossSales(grossSales);
        setCommissionRate(commissionRate);            
    }

    void CommissionEmployee::setCommissionRate (double commissionRate) {
        _commissionRate = (commissionRate > 0.0 && commissionRate < 1.0 ) ? commissionRate : 0.0;
    }

    void CommissionEmployee::setGrossSales (double grossSales) {
        _grossSales = (grossSales < 0.0 ) ? 0.0 : grossSales;
    }

    double CommissionEmployee::getCommissionRate () const {return _commissionRate;}
    double CommissionEmployee::getGrossSales () const {return _grossSales;}

    double CommissionEmployee::earnings() const {
        return getCommissionRate() * getGrossSales();
    }

    void CommissionEmployee::print() const {
        cout << "commission employee: ";
        Employee::print(); // code reuse
        cout << "\ngross sales: " << getGrossSales() << "; commission rate: " << getCommissionRate();    }   
}

File: cse241-object-oriented-programming/studies/java/inheritance/employee/CommissionEmployee.java
package employee;

public class CommissionEmployee extends Employee {
    protected double _grossSales;
    protected double _commissionRate;

    public CommissionEmployee (String fname, String lname, String ssn, double sales, double rate) {
        super(fname, lname, ssn);
        setGrossSales(sales); 
        setCommissionRate(rate); 
    }

    public void setGrossSales (double sales) {_grossSales = (sales < 0.0) ? 0.0 : sales;}

    public void setCommissionRate (double rate) {_commissionRate = (0.0 < rate && rate < 1.0) ? rate : 0.0;}

    public double grossSales () {return _grossSales;}

    public double commissionRate () {return _commissionRate;}

    public double earnings () {return commissionRate() * grossSales();}

    public String toString () {
        return String.format("%s: %s\n%s: $%,.2f; %s: %.2f", 
            "commission employee", super.toString(), 
            "gross sales", grossSales(),
            "commission rate", commissionRate() );
    }
}
